src/controller_examples/rabbitmq_controller/proof/liveness/liveness.rs

```diff
@@ -1894,4 +2803,45 @@ proof fn lemma_server_config_map_is_stable(
     leads_to_stable_temp(spec, lift_action(stronger_next), p, lift_state(post));
 }
 
+proof fn lemma_stateful_set_is_stable(
+    spec: TempPred<RMQCluster>, rabbitmq: RabbitmqClusterView, p: TempPred<RMQCluster>
+)
+    requires
+        spec.entails(p.leads_to(lift_state(current_stateful_set_matches(rabbitmq)))),
+        spec.entails(always(lift_action(RMQCluster::next()))),
+        spec.entails(always(lift_state(helper_invariants::no_delete_sts_req_is_in_flight(rabbitmq.object_ref())))),
+        spec.entails(always(lift_state(helper_invariants::every_update_sts_req_does_the_same(rabbitmq)))),
+        spec.entails(always(lift_state(helper_invariants::stateful_set_has_no_finalizers_or_timestamp_and_only_has_controller_owner_ref(rabbitmq)))),
```

Comment:
IIUC this precondition is used to ensure k8s API does not delete the object when handling update requests (the delete-during-update path). If so, could you document this a bit (just to remind ourselves in the future)?

Reply:
The precondition is an umbrella of several requirements. No timestamp is enough for showing that the object won't be deleted during update; no finalizers is added because this will make it easier to prove no timestamp without adding extra invariants; only having controller owner ref seems useless here, but it's just more convenient to put these invariants together and prove them all at once.